### ISSUE TITLE
Meeting keeps crashing on iOS Safari and Chrome

### ISSUE THREAD

**Issue by jowe13** [aws/amazon-chime-sdk-js#1309]
**Platform**
 - OS: iOS
 - Browser: Safari & Chrome
 - Browser Version: iOS 14.4.2, Chrome: latest from App Store (90.0.4430.78)
 - Amazon Chime SDK JS Version: 2.9.0

Hey guys,

we are trying to integrate Chime in our client React App. So far everything is working fine and as expected.
We are not facing any issues on Windows, Mac or Android. All browsers on these platforms are working just fine.

When trying to join a meeting we have the following scenarios:

1. We can force all attendees to a audio only meeting
2. Its up to the user to choose wether or not to use his/her camera.

Scenario 1: Everything is working fine.

In scenario two we got the following problems:

1. The user needs veeery long in order to join the meeting (sometimes up to a minute) when having camera and microphone activated when joining the meeting.
2. Once joining is done: All other attendees can't see but hear the iOS attendee. The iOS attendee can't see and can't here anyone. After a short time the whole page collapses and reloads. With the indication that an error happened repeatedly.
3. When joining with active camera and not selected microphone (is both selected before actually joining the meeting) everything works fine up to the point where the attendee selects his microphone while already in the meeting. Then the same error happens.
4. When joining without camera and selected default iPhone/iPad microphone the join time is much shorter then in all other cases but the same errors happens after a few seconds.

I don't get any chime or js errors in my js console. I tried monitoring iPhone/iPad logs with XCode developer console and got the following error appearing a few hundred times (again: only for iOS users).

<img width="750" alt="Bildschirmfoto 2021-05-21 um 14 04 57" src="https://user-images.githubusercontent.com/38559352/119134560-93c71000-ba3d-11eb-8d09-0cdb17509ab4.png">

Anyone has an advice on how to handle this problem? Or even why this happens on iOS only? Any help is much appreciated.

**Comment by devalevenkatesh** [comment]
Hi, thanks for reporting the issue.

To better understand the issue could you provide the following information?

1. **Meeting IDs and attendee IDs:** You can provide one sample meeting ID and attendee ID who experiences this issue.
2. **INFO-level browser logs of iOS users**: You can create an instance of the `ConsoleLogger` with `LogLevel.INFO`. For iOS Safari logs, please follow the instructions in [Enabling Web Inspector in the WebKit official website](https://webkit.org/web-inspector/enabling-web-inspector/).
3. The logs will help us understand the issue better on what is happening when you are selecting microphone and is there some reconnection issue going here. Since, the logs in the attached screenshot are not from JS SDK but looks like from iOS side, we wont be able to check the JS SDK further, but, was the user facing internet connectivity issue here?

Below is a snippet to add INFO level console logs to your app.
```js
import {
  ConsoleLogger,
  DefaultDeviceController,
  DefaultMeetingSession,
  LogLevel,
  MeetingSessionConfiguration,
} from 'amazon-chime-sdk-js';
const logger = new ConsoleLogger('ChimeMeetingLogs', LogLevel.INFO);
const deviceController = new DefaultDeviceController(logger);
const configuration = new MeetingSessionConfiguration(meeting, attendee);
const meetingSession = new DefaultMeetingSession(configuration, logger, deviceController);
```

Once, you add `ConsoleLogger` you can follow step 2 to get the logs in Desktop safari and then copy paste or attach those files here.

**Comment by jowe13** [comment]
Hey, thanks for your help.

MeetingId: beb55a01-d894-4571-ad14-b2a8cf8522c6
AttendeeId iOS: f239185e-9179-b2f5-9a97-dd402c2bbfc9
AttendeeId MacOS: be951d98-9959-3ade-0c18-1a98c98cce90

[chimelogs.txt](https://github.com/aws/amazon-chime-sdk-js/files/6537412/chimelogs.txt)

**Comment by devalevenkatesh** [comment]
Thanks for providing the info.

I checked the attached logs and looks like they are not complete. From the MeetingId you provided, I see the meeting as ended from server side, but, do not see corresponding browser logs which show whether the attendees left the meeting. So, is your application calling `audioVideo.stop()`?

Also, below two lines appear frequently in the logs:
```
DefaultDeviceController video dimension 960 x 540
Supported Constraints in this browser
```

JS SDK logs these lines in the `DeviceController` constructor [here](https://github.com/aws/amazon-chime-sdk-js/blob/22b82e54beaa1aca6bfb583903edf22f31f86e4e/src/devicecontroller/DefaultDeviceController.ts#L182) and [here](https://github.com/aws/amazon-chime-sdk-js/blob/22b82e54beaa1aca6bfb583903edf22f31f86e4e/src/devicecontroller/DefaultDeviceController.ts#L189) respectively. So, suspecting whether your app is either creating `DeviceController` more often. Is that the case? You should be creating the `DeviceController` only once.

We recommend you to check our JS SDK [meeting demo](https://github.com/aws/amazon-chime-sdk-js/tree/master/demos/browser) as well as our React component library [demo](https://github.com/aws/amazon-chime-sdk-component-library-react/tree/master/demo/meeting) for reference implementation. Can you please go through them once and let us know if you are still facing the same issue?

Summarizing questions from above initial checks:
1. Are the logs complete?
2. Does your application call `audioVideo.stop()`?
3. Is the `DeviceController` getting created more than once?

**Comment by jowe13** [comment]
Hi,

1. Yes the logs are 100% complete.
2. No my application or more like Safari just crashes.
3. Indeed that was a mistake and DeviceController was called multiple times. But that does not seem to fix this issue.

Please see new log which was taken after the DeviceController fix:

MeetingId: dcafe769-e12b-40d2-b165-e14e0ab5ae54
Attendee Id iOS: 2432d66c-0ba0-6607-414f-e279c28314ca
Attendee Id: macOS: f8edaac1-f33d-01e1-d1de-d4bec3eb8b19

[chimelogsupdated.txt](https://github.com/aws/amazon-chime-sdk-js/files/6544570/chimelogsupdated.txt)

Thanks for your help!

**Comment by devalevenkatesh** [comment]
1. The logs do not indicate any errors which show JS SDK crashed, so it may be problem in the way and time the APIs are called in your app.

2. I see below warning appearing in the logs:
```
[Warning] 2021-05-26T07:15:52.732Z [WARN] MyLogger - no transition found from NotConnected with Update (2.chunk.js, line 35173)
[Info] 2021-05-26T07:15:52.732Z [INFO] MyLogger - API/DefaultAudioVideoFacade/dcafe769-e12b-40d2-b165-e14e0ab5ae54/2432d66c-0ba0-6607-414f-e279c28314ca/startLocalVideoTile null -> 1 (2.chunk.js, line 35181)

```

This log is showing that JS SDK got an 'update' transition but since it was in 'NotConnected' state, it did not do anything.

I suspect, this is due to your app calling `startLocalVideoTile` before JS SDK is in 'Connected' state. The next log line indicates your app called `startLocalVideoTile`. The `startLocalVideoTile` call basically tries to turn your video on and send an update so that your app can bind the video stream to a video element for viewing. This will work if the JS SDK 'update' is triggered only after it is transitioned to 'Connected' state. Thus, this update is ignored as explained above as the JS SDK was in 'NotConnected' state.

3. In the startup process, JS SDK runs `ReceiveVideoStreamIndexTask` which checks whether it can receive new attendees videos or not and then sends an 'update'. Since, Step 3 would have been the same steps on the remote attendee, for the local attendee `ReceiveVideoStreamIndexTask` found that there is new videos to subscribe to. But, again that 'update' is deferred as the JS SDK is still in 'Connecting' state and not in 'Connected' state.
The 'update' will be deferred until later and performed once JS SDK goes from 'Connecting' -> 'Connected' state.
```
[Info] 2021-05-26T07:15:54.182Z [INFO] MyLogger - running task Signaling/ReceiveVideoStreamIndexTask (2.chunk.js, line 35181)
[Info] 2021-05-26T07:15:54.183Z [INFO] MyLogger - should resubscribe: true (downlink: true uplink: false) (2.chunk.js, line 35181)
[Info] 2021-05-26T07:15:54.183Z [INFO] MyLogger - trigger resubscribe for up=false down=true; videosToReceive=[1] captureParams={"cameraWidth":640,"cameraHeight":384,"cameraFrameRate":15,"maxEncodeBitrateKbps":1400,"isSimulcast":false} (2.chunk.js, line 35181)
[Info] 2021-05-26T07:15:54.183Z [INFO] MyLogger - deferring transition from Connecting with Update (2.chunk.js, line 35181)
```

4. For the long join time, this same task [`SetRemoteDescriptionTask`](https://github.com/aws/amazon-chime-sdk-js/blob/master/src/task/SetRemoteDescriptionTask.ts#L14) took almost 6s to complete. This task is basically WebRTC peer dependent. I suspect this may be the same case in any WebRTC based app from your test device facing this issue. Request you to try [appr.tc](https://appr.tc/) which is the WebRTC's test website.

```
[Info] 2021-05-26T07:15:59.973Z [INFO] MyLogger - AudioVideoStart/dcafe769-e12b-40d2-b165-e14e0ab5ae54/2432d66c-0ba0-6607-414f-e279c28314ca/Timeout15000ms/Peer/SetRemoteDescriptionTask took 5663 ms (2.chunk.js, line 35181)
```


Can you please do the following and see whether this solves the issues?
1. For second scenario, check whether `audioVideo.start()` completes and then only allow users to start / stop video.
2. You can implement [`audioVideoDidStart`](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideoobserver.html#audiovideodidstart) method from [`AudioVideoObserver`](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideoobserver.html) to know whether the JS SDK is fully connected and `audioVideo.start()` has completed.
3. Try https://appr.tc/ and check whether you are facing similar delays.

Apart from the above observation in the logs I could not find any crash or error within JS SDK. Let us know if you have further questions.

**Comment by jowe13** [comment]
Hi,

apparently it had nothing to do with any SDK functions being called early.
It was an issue with the reusing of video elements. The problem is solved for me.
It is also as fast as on desktop now in terms of meeting join times.

Thanks for your time and patience.
